school/database.py

```diff
@@ -0,0 +1,514 @@
+    ),
+)
+
+
+class SubjectProgram(database.base):
+    __tablename__ = "school_dataset_subject_program"
+
+    subject_idx = Column(
+        ForeignKey("school_dataset_subject.idx", ondelete="CASCADE"),
+        primary_key=True,
+    )
+    program_idx = Column(
+        ForeignKey("school_dataset_program.idx", ondelete="CASCADE"),
+        primary_key=True,
+    )
+    year = Column(Integer, primary_key=True)
+    obligation = Column(String, primary_key=True)
+    subject = relationship("Subject", back_populates="programs")
+    program = relationship("Program", back_populates="subjects")
+
+    @staticmethod
+    def get(subject: Subject, program: Program, year: int, obligation: str):
+        query = (
+            session.query(SubjectProgram)
+            .filter_by(subject_idx=subject.idx)
+            .filter_by(program_idx=program.idx)
+            .filter_by(year=year)
+            .filter_by(obligation=obligation)
+            .one_or_none()
+        )
+
+        return query
+
+    @staticmethod
+    def add_relation(subject: Subject, program: Program, year: int, obligation: str):
+        relation = SubjectProgram(
+            subject_idx=subject.idx,
+            program_idx=program.idx,
+            year=year,
+            obligation=obligation,
+        )
+
+        session.add(relation)
+        session.commit()
+
+        return relation
+
+    def delete(self):
+        session.delete(self)
+        session.commit()
+
+
+class Teacher(database.base):
+    __tablename__ = "school_dataset_teacher"
+    __table_args__ = (
+        UniqueConstraint("school_id", "guild_id", name="school_id_guild_id_unique"),
+    )
+
+    idx = Column(Integer, primary_key=True)
+    school_id = Column(Integer)
+    guild_id = Column(BigInteger)
+    name = Column(String)
+    subjects = relationship(
+        "Subject",
+        secondary=teachers_subjects,
+        back_populates="teachers",
+    )
+    guaranted_subjects = relationship(
+        "Subject",
+        back_populates="guarantor",
+    )
+
+    _used_filter_generators = []
+    _is_used_checks = []
+
+    @staticmethod
+    def add_used_filter_generator(func: Callable):
+        """Add callable to list of functions which are called
+        to get filter of used teacher.
+
+        The function must return filter statement, for example:
+
+        `stmt = exists().where(Teacher.school_ud==MyDatabaseObject.foreign_teacher_id)`
+
+        """
+        if func is None:
+            return
+
+        if func not in Teacher._used_filter_generators:
+            Teacher._used_filter_generators.append(func)
+
+    @staticmethod
+    def add_is_used_check(func: Callable):
+        """Add callable to list of functions which are called
+        to check if teacher is in use somewhere.
+
+        The function must return Boolean.
+        """
+        if func is None:
+            return
+
+        if func not in Teacher._is_used_checks:
+            Teacher._is_used_checks.append(func)
```

Comment:
I plan to add warning to subject / teacher delete embed that the object is in use. Then if the object is deleted, the corresponding reviews are deleted too (or in case of guarantor it's set to NULL)